qmi/core/transport.py

```diff
@@ -1071,29 +1091,40 @@ def read_until(self, message_terminator: bytes, timeout: Optional[float]) -> byt
             else:
                 raise QMI_InstrumentException("Error attempting to read bytes until {!r} char received".format(
                     message_terminator)) from err
+
         finally:
             # Revert terminator and timeout to previous value
             self._safe_instr.term_char = old_term_char
             self._safe_instr.timeout = old_timeout
 
-        return data
+        ret = self._read_buffer
+        self._read_buffer = bytes()
+        return ret
+
+    def read_until_timeout(self, nbytes: int, timeout: float):
+        return self.read(nbytes, timeout)
```

Comment:
Right, QMI_TimeoutException should be caught indeed. Then it can return whatever was read in the buffer until that exception and clear the buffer. In any other exception, if something was read in, it'll stay in the `_read_buffer`.